### ISSUE TITLE
[BUG] self.generator uninitialized

### ISSUE THREAD

**Issue by lucyknada** [theroyallab/tabbyAPI#73]
**Is your feature request related to a problem? Please describe.**
Tabby is known to not handle multiple requests well, if you try; it'll break until you restart it, one example being: [self.generator uninitialized](https://github.com/theroyallab/tabbyAPI/blob/fc857893eea28c895991b4b272c8d5a8e03539ac/backends/exllamav2/model.py#L715)

**Describe the solution you'd like**
A queue system that delays other requests and has workers process N-amount of queue items in parallel, if impossible with the codebase to process in parallel, having it do it one by one is fine too until parallel processing could be implemented

**Describe alternatives you've considered**
Writing a proxy that shields tabby and implements the queue infront of it, waiting each time for tabby to return a response before continuing, which... is an awkward bandaid solution

**Comment by bdashore3** [comment]
Tabby already has a sequential queuing system that works in the same way as other backends like text-generation-webui. Parallel batching is planned for future work.

In your case, the Generator wasn't being initialized properly which is usually due to a possible issue with model loading. But, there's not much information provided for me to make definite conclusions. I'd need more information to proceed, preferably in the [Support Server](https://discord.gg/sYQxnuD7Fj) as issues is for code-sided bugs.

In addition, multi-user queuing is moreso meant for small scales. For larger scale or parallel requests, I'd suggest looking into engines such as [Aphrodite Engine](https://github.com/PygmalionAI/aphrodite-engine) or [VLLM](https://github.com/vllm-project/vllm)

**Comment by lucyknada** [comment]
it probably allows to request completions before a model was finished loading, that could be fixed on its own, though I assume other operations might shoot through with the same bug, would a global queue instead be preferable without having to drill into all possible scenario permutations?

```javascript
const express = require("express");
const queue = require("express-queue");
const proxy = require("express-http-proxy");
const app = express();
app.use(queue({ activeLimit: 1, queuedLimit: -1 }));
app.use("/", proxy("127.0.0.1:5000"));
app.listen(5500);
```

that bandaid proxy simulates that behavior I'm talking about, where all requests are queued up as they come in and processed one at a time, shielding tabby from those non-exclusive ops.

**Comment by bdashore3** [comment]
Thanks for the extra info! Yes, I was able to reproduce this behavior. While a global queue is a good idea, I'd like to have the queue only run for completions and add checks instead on generations. Marking as a bug.

**Comment by bdashore3** [comment]
Update: Please try the [API concurrency branch](https://github.com/theroyallab/tabbyAPI/tree/api-concurrency) which should fix these issues.

**Comment by lucyknada** [comment]
while that kind of solves the bug of generator being undefined;

the model load is still non-blocking, that means if you are executing the load while something else is requesting new completions you'll get spammed by:

```
ERROR: NoneType: None
ERROR: 
ERROR: No models are currently loaded.
ERROR: NoneType: None
ERROR: 
ERROR: No models are currently loaded.
ERROR: NoneType: None
ERROR: 
ERROR: No models are currently loaded.
ERROR: NoneType: None
```

until tabby seemingly just crashes / stops trying to load the model and on next request you just get:

`400 {"detail":"No models are currently loaded."}`

the forward queue from above blocks all subsequent requests until the load is done, so it avoids that scenario automatically

**Comment by AlpinDale** [comment]
I would throw out a "No models are currently loaded" error, but if the model loading process is fast enough, I believe you can get away with queuing the request.

**Comment by bdashore3** [comment]
Should be resolved in 165cc6fc2d351010443f26cdd0e48abb4296323f and below.

Please track #77 for further updates. Closing this issue.
